### PR TITLE
feat(fsdp_tp): async DCP checkpointing + backgrounded fan-out + meta-device init (large-model checkpoint/restart, validated on Sunspot)

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Makes `ezpz.examples.fsdp_tp` checkpoint **asynchronously** and scale to **large
models** — two capabilities that compound: async removes the per-save training
stall, and meta-device init lets the model that most needs cheap checkpointing
(a 20B, 232 GB checkpoint) actually run. Builds on the sync checkpointing in
#196 (v0.23.0). **Every result below is measured on Sunspot XPU.**

The PR grew past its original "async DCP" scope into a coherent
checkpoint-and-scale arc; this description reflects the whole thing.

## What's here

**1. Async DCP checkpointing** (`--async-ckpt`, `--ckpt-stage-dir`)
`dcp.async_save` stages state to CPU synchronously (safe to keep training),
writes shards to node-local `/tmp` on a background thread, then fans them out
to the durable `--ckpt-dir` on shared FS. Resume is always from the durable
copy — `/tmp` is a staging tier, never resumable on its own (enforced by a
test).

**2. Backgrounded fan-out** (`start_fanout` / `try_finalize_if_ready` /
`finalize_fanout`)
The fan-out copy is the expensive half of an async save (5–20 s at scale). It's
pure per-rank file I/O with **no collectives**, so it runs on a background
worker; a cheap per-step **MPI** readiness probe (MPI's own communicator, *not*
the xccl training PG — so it can't cross-match gradient collectives and
deadlock) stamps the durable `.complete` marker the moment all ranks' copies
land. This overlaps the copy with training instead of blocking the loop, and
keeps the durability window to ~1 copy-duration.

**3. Meta-device init** (`--meta-init {auto,on,off}`, default `auto`)
Large models OOM'd at build: the example moved the full dense model onto one GPU
before FSDP sharded it, capping model size at ~2–8B regardless of node count.
Meta-init (torchtitan pattern) builds on `meta` → shards → `to_empty` →
`init_weights(buffer_device=device)`, so only each rank's shard is ever
materialized. `auto` enables it for native models ≳6B and keeps small models on
the exact dense path (sharded DTensor init isn't bit-for-bit with dense, so
small models stay reproducible); HF `from_pretrained` never uses it.

**4. Correct throughput metrics under TP**
`train/tps_per_gpu` (÷ world_size), `train/tflops` and `train/mfu` (÷ tp) were
over-counting by the parallelism degree; fixed with regression tests. New save
metrics: `ckpt_save_seconds`, `ckpt_stage_seconds`, `ckpt_drain_seconds`.

## Measured results (Sunspot)

| model | ckpt size | sync save stall | async (stage+drain) | speedup |
|---|---:|---:|---:|---:|
| agpt-2b | 23 GB | 3.75 s | 1.05 s | ~3.6× |
| agpt-20b | 232 GB | 23.6 s | 5.4 s | ~4.4× |

The async win **scales with checkpoint size** — the bigger the model, the more
the fan-out is worth hiding. At 20b, async also finishes on wall-clock parity
(both resume from identical checkpoints; lost steps matched). Restart cost is
dominated by `dcp.load` (~40 s at 2b, ~60 s at 20b).

**Meta-init fit test:** agpt-20b peak memory OOM (>64 GB/tile) → **14 GB/tile**
with `--meta-init on`; `--meta-init off` reproduces the OOM at the exact same
`model.to(device)` line (control confirms the flag is the lever). Validated
across 24- and 48-rank runs with no cross-thread deadlock and clean DCP resume
from a 232 GB checkpoint.

## The hard parts (all found + fixed on real hardware)

`dcp.async_save` on XPU surfaced platform issues the single-process CI tests
structurally cannot:

1. **`async_save` requires a CPU backend** — XPU's default PG is xccl-only.
   Fixed with a **composite** backend `xpu:xccl,cpu:gloo` (accel-first,
   torchtitan pattern).
2. **`device_id` suppressed the CPU backend** — `_setup_ddp` now drops
   `device_id` only for composite backends; normal single-backend runs keep it
   (load-bearing FSDP2-deadlock guard).
3. **`drain` fan-out raced** — each rank now copies only its own
   `__<rank>_*.distcp` with per-rank temp names.
4. **Untimed drain hid the real async cost** — the naive `ckpt_stage_seconds`
   made async *look* 12× faster while it was actually slower; timing the drain
   exposed it, and backgrounding the fan-out then made async genuinely win.
   Documented honestly in the guide.

## Tests
- `test_checkpoint.py`: async round-trip, stage-only-not-resumable guard,
  drain/start/finalize idempotency + none-safety, previous-checkpoint-durable
  window, background copy round-trip.
- `test_fsdp_tp_meta_init.py`: tier estimate/resolution (2b dense vs 20b meta,
  env threshold, HF-forces-dense) + real meta build → to_empty → init_weights.
- `test_fsdp_tp_dp_degrees.py`: tps_per_gpu / tflops / mfu ÷-degree regression.
- 77 tests pass locally; ruff clean on changed files; CI green.

## Docs
`docs/guides/checkpoint-restart.md` — full guide with the 2b + 20b plots and the
honest before→after of the drain accounting. `experiments/checkpoint-restart/`
— reproducible driver (env-overridable for larger models) + plotters.

## Test plan
- [x] Local: 77 tests + ruff clean
- [x] Sunspot: composite backend, async stage + backgrounded fan-out + resume,
      meta-init 20b fit (+ OOM control), 2b & 20b sync-vs-async experiments

### CODE REVIEW

File: src/ezpz/examples/_checkpoint.py [hunk 1]
```diff
@@ -117,18 +135,139 @@ def save_checkpoint(
...
+    The staged shards are being written to ``stage_path`` (node-local) by a
+    DCP background thread; ``future`` resolves when that disk write finishes.
+    :func:`drain` then fans the shards out to the durable ``final_path`` on
+    shared FS and stamps the completion marker there. Until drained, the
+    durable checkpoint does NOT exist / is not marked complete.
+    """
+
+    future: Any
+    step: int
+    meta: Optional[dict]
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/_checkpoint.py]: Fixed in 4a41ca8 — PendingCheckpoint.meta is now Optional[dict[str, Any]], matching save_checkpoint_async.

File: tests/test_checkpoint.py [hunk 2]
```diff
@@ -203,3 +203,78 @@ def test_load_returns_none_when_empty(self, tmp_path):
...
+    def test_drain_is_idempotent_and_none_safe(self, tmp_path):
+        torch = _torch_or_skip()
+        _init_single_rank_pg(torch)
+        m = _import()
+        assert m.drain(None) is None
+        model, opt = _Tiny.build(torch)
+        opt.step()
+        pending = m.save_checkpoint_async(
+            tmp_path / "c", tmp_path / "s", 5, model, opt, meta={}
+        )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_checkpoint.py:277]: Strengthened in 4a41ca8 — the drain idempotency test now also asserts (p1/".complete").exists() and that the stage dir (tmp_path/"s"/"step-5") was reclaimed.

File: src/ezpz/examples/_checkpoint.py [hunk 3]
```diff
@@ -117,18 +135,139 @@ def save_checkpoint(
...
-        # safe to resume from" contract that latest_checkpoint() relies on.
-        (out / _COMPLETE_MARKER).write_text("")
+        _write_meta_and_marker(out, step, meta)
     _barrier()
     if is_rank0:  # gate per-rank log spam (AGENTS.md)
         logger.info("saved checkpoint: %s", out)
     return out
 
 
+@dataclass
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/_checkpoint.py:161]: Thanks — I kept the current PendingCheckpoint/drain shape for this PR: behavior is correct and hardware-validated (Sunspot job 12471713), and the Optional handling doubles as the None-safe/idempotent contract the tests lock. Noting the encapsulation refactor as a follow-up rather than churning validated code.

File: src/ezpz/examples/fsdp_tp.py [hunk 4]
```diff
@@ -3159,23 +3237,58 @@ def train(
...
+                if args.async_ckpt:
+                    from ezpz.examples._checkpoint import (
+                        drain,
+                        save_checkpoint_async,
+                    )
+
+                    # Only one async save may be in flight (DCP constraint) —
+                    # finish the previous one before starting the next. At
+                    # save_interval >> save time this is effectively free.
+                    drain(pending_ckpt)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Good catch (P1) — fixed in 4a41ca8. drain() now runs at the START of the next step, not deferred to the next save interval, so the staged-but-not-durable window is ~1 step instead of ~1 interval. The background write has usually finished by then, so drain rarely blocks; a crash loses at most ~1 interval.

File: src/ezpz/examples/fsdp_tp.py [hunk 5]
```diff
@@ -1699,9 +1723,48 @@ def parse_args(argv: Optional[list[str]] = None):
...
+    ``setup_torch``/``init_process_group`` — and sets ``TORCH_BACKEND``, which
+    ``ezpz.distributed.get_torch_backend`` honors as an override.
+
+    Respects a user-set ``TORCH_BACKEND`` (only fills it when unset) and is a
+    no-op on CPU-only runs (backend is already gloo).
+    """
+    if not getattr(args, "async_ckpt", False):
+        return
+    if os.environ.get("TORCH_BACKEND"):
+        return  # user override wins
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Fixed in 4a41ca8 — an override lacking a CPU backend (e.g. TORCH_BACKEND=xccl) is now augmented to "<override>,cpu:gloo" instead of returning early; composites that already contain cpu:/gloo are left untouched. Verified xccl -> xccl,cpu:gloo.

File: src/ezpz/examples/fsdp_tp.py [hunk 6]
```diff
@@ -1699,9 +1723,48 @@ def parse_args(argv: Optional[list[str]] = None):
...
+        else:
+            return  # CPU-only: default gloo already has a CPU backend
+        # Order matches torchtitan (distributed/utils.py): accelerator FIRST,
+        # then cpu:gloo. The accel backend stays primary; cpu:gloo is added so
+        # dcp.async_save's background CPU-thread collective has a backend.
+        os.environ["TORCH_BACKEND"] = f"{accel},cpu:gloo"
+        logger.info(
+            "async-ckpt: set TORCH_BACKEND=%s so async_save has a CPU backend",
+            os.environ["TORCH_BACKEND"],
+        )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Fixed in 4a41ca8 — the backend-setup log is now gated to env local-rank 0 (PALS_LOCAL_RANKID/PMIX_LOCAL_RANK/LOCAL_RANK), so it fires once per host, not once per rank.

File: src/ezpz/examples/fsdp_tp.py [hunk 7]
```diff
@@ -1699,9 +1723,48 @@ def parse_args(argv: Optional[list[str]] = None):
...
 
+def _maybe_enable_cpu_backend_for_async_ckpt(args: argparse.Namespace) -> None:
+    """Register a CPU (gloo) backend when --async-ckpt is set.
+
+    ``dcp.async_save`` asserts the process group includes a CPU backend (it
+    stages/writes from a background CPU thread). On an accelerator the default
+    PG is xccl/nccl-only, so we must init with a COMPOSITE backend string
+    (e.g. ``cpu:gloo,xpu:xccl``). This runs in parse_args — BEFORE
+    ``setup_torch``/``init_process_group`` — and sets ``TORCH_BACKEND``, which
+    ``ezpz.distributed.get_torch_backend`` honors as an override.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:1758]: Fixed in 20c6351 — docstring composite example now matches the code (accelerator-first "xpu:xccl,cpu:gloo").

File: src/ezpz/examples/fsdp_tp.py [hunk 8]
```diff
@@ -1699,9 +1723,48 @@ def parse_args(argv: Optional[list[str]] = None):
...
+        return  # user override wins
+    try:
+        import torch
+
+        if torch.cuda.is_available():
+            accel = "cuda:nccl"
+        elif torch.xpu.is_available():
+            accel = "xpu:xccl"
+        else:
+            return  # CPU-only: default gloo already has a CPU backend
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:1800]: Fixed in 20c6351 — the XPU probe is now guarded with hasattr(torch, "xpu"), so CPU-only/CUDA-only/ROCm builds no longer raise AttributeError (and --async-ckpt on CPU-only no longer emits a false warning).

File: docs/guides/checkpoint-restart.md [hunk 9]
```diff
@@ -33,6 +33,33 @@ ezpz launch --auto-retry --np <N> -- \
...
+    --ckpt-dir /shared/ckpts --async-ckpt \
+    --ckpt-stage-dir /tmp/ezpz-ckpt --save-interval 100
+```
+
+`--async-ckpt` uses `dcp.async_save`: the state dict is staged to CPU memory
+**synchronously** (so it's safe to keep training the instant the call
+returns), then written to fast **node-local** `--ckpt-stage-dir` (default
+`/tmp/ezpz-ckpt-<jobid>`) by a background thread, and finally **fanned out**
+to the durable `--ckpt-dir` on shared FS. The only cost on the training
+thread is the short staging copy (logged as `ckpt_stage_seconds`).
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/checkpoint-restart.md]: Fixed in 20c6351 — the guide now names the fully-qualified metric train/ckpt_stage_seconds.

File: src/ezpz/examples/fsdp_tp.py [hunk 10]
```diff
@@ -3159,23 +3237,58 @@ def train(
...
+                        meta=_ckpt_meta,
+                    )
+                    # Caller-thread stall = staging time (the disk write +
+                    # fan-out happen off-thread). Compare to the sync save cost.
+                    if ezpz.get_rank() == 0:
+                        logger.info(
+                            "ckpt_stage_seconds=%.4f (async stage @ step %d)",
+                            perf_counter() - _t_stage,
+                            global_step,
+                        )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Fixed in 20c6351 — it is now a real tracked metric via history.tracker.log({"train/ckpt_stage_seconds": ...}) (lands in History/W&B/JSONL), in addition to the log line.
